src/Graph.ts

```diff
@@ -227,99 +267,254 @@ export default class Graph {
 			.then(entryModule => {
 				timeEnd('phase 1');
 
-				// Phase 2 - linking. We populate the module dependency links
-				// including linking binding references between modules. We also
+				// Phase 2 - linking. We populate the module dependency links and
 				// determine the topological execution order for the bundle
 				timeStart('phase 2');
 
 				this.link();
-				const { orderedModules, dynamicImports } = this.analyseExecution(entryModule);
-				dynamicImports.forEach(dynamicImportModule => {
-					if (orderedModules.indexOf(dynamicImportModule) === -1)
-						orderedModules.push(dynamicImportModule);
-				});
-				const bundle = new Bundle(this, orderedModules);
+
+				const { orderedModules, dynamicImports, hasCycles } = this.analyseExecution([entryModule]);
+				if (hasCycles) {
+					this.warnCycle(entryModule, orderedModules);
+				}
 
 				timeEnd('phase 2');
 
 				// Phase 3 – marking. We include all statements that should be included
 				timeStart('phase 3');
 
-				// mark all export statements for the entry module and dynamic import modules
-				bundle.bind();
-
 				entryModule.markExports();
+
 				dynamicImports.forEach(dynamicImportModule => {
-					dynamicImportModule.markExports();
+					if (entryModule !== dynamicImportModule)
+						dynamicImportModule.markExports();
+					// all dynamic import modules inlined for single-file build
 					dynamicImportModule.namespace().includeVariable();
 				});
 
 				// only include statements that should appear in the bundle
-				bundle.includeMarked(this.treeshake);
+				this.includeMarked(orderedModules);
 
 				// check for unused external imports
 				this.externalModules.forEach(module => module.warnUnusedImports());
 
 				timeEnd('phase 3');
 
-				// Phase 4 – we ensure that names are deconflicted throughout the bundle
+				// Phase 4 – we construct the chunk itself, generating its import and export facades
+				timeStart('phase 4');
+
+				// generate the imports and exports for the output chunk file
+				const chunk = new Chunk(this, entryModule.id, orderedModules);
+				chunk.collectDependencies();
+				chunk.generateImports();
+				chunk.generateEntryExports(entryModule);
+
+				timeEnd('phase 4');
+
+				return chunk;
+			});
+	}
+
+	buildChunks (entryModuleIds: string[]): Promise<{ [name: string]: Chunk }> {
```

Comment:
At it's made an improvement on the `buildSingle` function though from where it was before as the main build method.

I'm still weary of over-abstracting though, it can be just as much or even more of a maintenance issue having to jump around a file to work out exactly what a method is doing. Seeing the low-level details in one place is a powerful way to understand something even if it does mean you have to stare at it for longer than 5 minutes :P

Reply:
Also the idea is the code comments provide the high-levels... reading through the comments gives an algorithmic overview, better than some clever method names might.